### PR TITLE
Fix dice score when zero overlap between preds and target

### LINKED ISSUE #2850: `DiceScore` yields 1.0 scores when using `average="none"` for samples where class is not present
## 🐛 Bug

The current implementation of `DiceScore` yields scores of 1.0 for samples that don't contain a particular class when calculating class-wise metrics via `average="none"`.

This leads to very high dice scores, particularly for rare classes, as samples where the class is not present will push the metric towards 1.0. It also makes dice scores among classes incomparable, unless the val/test dataset is balanced.

### To Reproduce

The following code snipped shows an example where for 1000 samples, the first class is only present in the first sample. Even though the prediction on that first sample is wrong, it yields a close to perfect class dice score of 99.9%, because for the 999 samples where the class is absent, a score of 1.0 is used.

<details>
  <summary>Code sample</summary>

```python
import torch
from torchmetrics.segmentation import DiceScore

N_SAMPLES = 1000
N_CLASSES = 3

target = torch.full((N_SAMPLES, N_CLASSES, 128, 128), 0, dtype=torch.int8)
preds = torch.full((N_SAMPLES, N_CLASSES, 128, 128), 0, dtype=torch.int8)

# The first class is only present in the first sample, but prediction is wrong
target[0, 0] = 1 

dice = DiceScore(num_classes=3, average='none', include_background=True)

print(dice(preds, target))
# prints: tensor([0.9990, 1.0000, 1.0000])
```

</details>

### Expected behavior
The above code snipped should return `[0.0, nan, nan]` (nan for second & third classes because they are not present in any of the provided samples, therefore no meaningful score can be calculated).

### Environment

- TorchMetrics version (if build from source, add commit SHA): `1.6.0`
- Python & PyTorch Version (e.g., 1.0): Python 3.11.10 & `torch==2.3.0`
- Any other relevant information such as OS (e.g., Linux): macOS 15.1.1 (24B91)

### Additional context

- Very similar to issue #2846 for `GeneralizedDiceScore`, but there `0.0` scores for samples where class is missing are used.
- The `MONAI` library addresses this through the `ignore_empty` keyword argument, which is set to `True` by default (see [monai.metrics.DiceMetric](https://docs.monai.io/en/latest/metrics.html#mean-dice))

### LINKED ISSUE #2851: `DiceScore(average="weighted")` produces scores bigger than 1.0
## 🐛 Bug

Using `average=weighted` aggregation in the `DiceScore` produces scores bigger than 1.0, and therefore outside the expected value range of [0,1] for dice.

### To Reproduce

The following code sample returns a score of 2.0:

<details>
  <summary>Code sample</summary>

```python
import torch
from torchmetrics.segmentation import DiceScore

N_SAMPLES = 4
N_CLASSES = 3

target = torch.full((N_SAMPLES, N_CLASSES, 128, 128), 0, dtype=torch.int8)
preds = torch.full((N_SAMPLES, N_CLASSES, 128, 128), 0, dtype=torch.int8)

target[0, 0], preds[0, 0] = 1, 1

dice = DiceScore(num_classes=3, average='weighted', include_background=False)

print(dice(preds, target))
```

</details>

### Environment

- TorchMetrics version (if build from source, add commit SHA): `1.6.0`
- Python & PyTorch Version (e.g., 1.0): Python 3.11.10 & `torch==2.3.0`
- Any other relevant information such as OS (e.g., Linux): macOS 15.1.1 (24B91)

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #2851
Fixes #2850

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2860.org.readthedocs.build/en/2860/

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Comment by Borda** [comment]
@SkafteNicki lets finish this one :)
